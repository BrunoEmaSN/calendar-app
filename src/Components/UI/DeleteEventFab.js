import React from 'react'
import { useDispatch } from 'react-redux';
import { eventRemove } from '../../Store/Calendar/EventActions';

export const DeleteEventFab = () => {
    const dispatch = useDispatch()
    const handleClick = () => {
        dispatch( eventRemove() );
    }
    return (
        <button className="btn. btn-danger fab-danger" onClick={ handleClick }>
            <i className="fas fa-trash"></i>
            <span> Borrar Evento </span>
        </button>
    )
}