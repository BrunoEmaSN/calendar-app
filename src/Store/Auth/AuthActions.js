import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../../Helpers/fecth";
import { types } from "../../Types";
import { eventActiveCleaning } from "../Calendar/EventActions";

export const startLogin = (email, password) => {
    return async ( dispatch ) => {
        const resp = await fetchWithoutToken( 'auth/login', { email, password }, 'POST' );
        const body = await resp.json();

        if( body.ok ){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch( login({
                uid: body.uid,
                name: body.name
            }) )
        }
        else {
            Swal.fire('Error', body.msg, 'error');
        }
    }
}

export const startRegister = ( email, password, name ) => {
    return async ( dispatch ) => {
        const resp = await fetchWithToken( 'auth/', { name, email, password }, 'POST' );
        const body = await resp.json();

        if( body.ok ){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch( login({
                uid: body.uid,
                name: body.name
            }) );
        }
        else {
            Swal.fire('Error', body.msg, 'error');
        }
    }
}

const login = ( user ) => ({
    type: types.AuthLogin,
    payload: user
});

export const startLogout = () => {
    return (dispatch) => {
        localStorage.clear();
        dispatch( eventActiveCleaning() );
        dispatch( logout() );
    }
}

const logout = () => ({
    type: types.AuthLoguot
});

export const startChecking = () => {
    return async( dispatch ) => {
        const resp = await fetchWithToken( 'auth/reToken', 'POST' );
        const body = await resp.json();

        if( body.ok ){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch( login({
                uid: body.uid,
                name: body.name
            }) );
        }
        else {
            dispatch( checkingFinish() );
        }
    }
}

const checkingFinish = () => ({
    type: types.AuthCheckingFinish
});