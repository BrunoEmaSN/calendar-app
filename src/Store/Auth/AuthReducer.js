import { types } from "../../Types";

const initialState = {
    checking: true
}

export const authReducer = ( state = initialState, actions ) => {
    switch (actions.type) {
        case types.AuthLogin:
            return {
                ...state,
                checking: false,
                ...actions.payload
            };

        case types.AuthCheckingFinish:
            return {
                ...state,
                checking: false
            }
        
        case types.AuthLoguot:
            return {
                checking: false
            };

        default:
            return state;
    }
}