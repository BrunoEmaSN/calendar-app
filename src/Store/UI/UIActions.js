import { types } from "../../Types";

export const openModal = () => ({ type: types.UIOpenModal });

export const closeModal = () => ({ type: types.UICloseModal });

export const setError = ( err ) => ({
    type: types.uiSetError,
    payload: err
});

export const rmError = () => ({
    type: types.uiRmError,
    payload: null
});