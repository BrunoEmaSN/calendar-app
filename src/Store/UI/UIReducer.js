import { types } from "../../Types";

const initialState = {
    modalOpen: false,
    msgError: null
}

export const UIReducer = ( state = initialState, actions ) => {
    switch ( actions.type ) {
        case types.UIOpenModal:
            return {
                ...state,
                modalOpen: true
            };

        case types.UICloseModal:
            return {
                ...state,
                modalOpen: false
            };
        
        case types.uiSetError:
            return {
                ...state,
                msgError: actions.payload
            };
            
        case types.uiRmError:
            return {
                ...state,
                msgError: null
            };
    
        default:
            return state;
    }
}