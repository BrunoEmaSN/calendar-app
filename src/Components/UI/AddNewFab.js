import React from 'react'
import { useDispatch } from 'react-redux';
import { openModal } from '../../Store/UI/UIActions';

export const AddNewFab = () => {
    const dispatch = useDispatch()
    const handleClick = () => {
        dispatch( openModal() );
    }
    return (
        <button className="btn. btn-primary fab" onClick={ handleClick }>
            <i className="fas fa-plus"></i>
        </button>
    )
}
