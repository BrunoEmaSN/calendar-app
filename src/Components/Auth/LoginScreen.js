import React from 'react';
import Swal from 'sweetalert2';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from '../../Hooks/useForm';
import { rmError, setError } from '../../Store/UI/UIActions';
import { startLogin, startRegister } from '../../Store/Auth/AuthActions';
import './login.css';

export const LoginScreen = () => {
    const dispatch = useDispatch();

    const { msgError } = useSelector( state => state.UI );

    const [ formLoginValues, handleLoginInputChange ] = useForm({
        lEmail: '',
        lPassword: ''
    });
    
    const { lEmail, lPassword } = formLoginValues;

    const [ formRegisterValues, handleRegisterInputChange ] = useForm({
        rName: '',
        rEmail: '',
        rPassword: '',
        rPasswordConfirm: ''
    });

    const { rName, rEmail, rPassword, rPasswordConfirm } = formRegisterValues;

    const isFormLoginValid = () => {
        if( !validator.isEmail( lEmail ) ) {
            dispatch( setError( 'Correo invalido' ) );
            return false;
        }
        
        if( lPassword.length < 6 ){
            dispatch( setError( 'La contraseña debe de contener minimo 6 caracteres' ) );
            return false;
        }

        dispatch( rmError() );
        return true;
    }

    const isFormRegisterValid = () => {
        if( rName.length < 3 ){
            dispatch( setError( 'El nombre debe de contener minimo 3 caracteres' ) );
        }
        if( !validator.isEmail( rEmail ) ) {
            dispatch( setError( 'Correo invalido' ) );
            return false;
        }
        
        if( rPassword.length < 6 ){
            dispatch( setError( 'La contraseña debe de contener minimo 6 caracteres' ) );
            return false;
        }

        if( !validator.equals(rPassword, rPasswordConfirm) ){
            dispatch( setError( 'Las contraseñas deben de ser iguales' ) );
            return false;
        }

        dispatch( rmError() );
        return true;
    }

    const handleLogin = ( e ) => {
        e.preventDefault();
        if( isFormLoginValid() ){
            dispatch( startLogin( lEmail, lPassword ) );
        }
        else {
            Swal.fire('Error', msgError, 'error');
        }
    }

    const handleRegister = ( e ) => {
        e.preventDefault();
        if( isFormRegisterValid() ){
            dispatch( startRegister( rEmail, rPassword, rName ) )
        }
        else {
            Swal.fire('Error', msgError, 'error');
        }
    }

    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Ingreso</h3>
                    <form onSubmit={ handleLogin }>
                        <div className="form-group">
                            <input 
                                type="text"
                                className="form-control"
                                placeholder="Correo"
                                name="lEmail"
                                value={ lEmail }
                                onChange={ handleLoginInputChange }
                                required
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contraseña"
                                name="lPassword"
                                value={ lPassword }
                                onChange={ handleLoginInputChange }
                                required
                            />
                        </div>
                        <div className="form-group">
                            <input 
                                type="submit"
                                className="btnSubmit"
                                value="Login"
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Registro</h3>
                    <form onSubmit={ handleRegister }>
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Nombre"
                                name="rName"
                                value={ rName }
                                onChange={ handleRegisterInputChange }
                                required
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Correo"
                                name="rEmail"
                                value={ rEmail }
                                onChange={ handleRegisterInputChange }
                                required
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contraseña"
                                name="rPassword"
                                value={ rPassword }
                                onChange={ handleRegisterInputChange }
                                required
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Repita la contraseña"
                                name="rPasswordConfirm"
                                value={ rPasswordConfirm }
                                onChange={ handleRegisterInputChange }
                                required
                            />
                        </div>

                        <div className="form-group">
                            <input 
                                type="submit" 
                                className="btnSubmit" 
                                value="Crear cuenta" />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}