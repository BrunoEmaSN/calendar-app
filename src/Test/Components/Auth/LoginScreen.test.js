import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';
import { LoginScreen } from '../../../Components/Auth/LoginScreen';
import { startLogin, startRegister } from '../../../Store/Auth/AuthActions';
import { rmError, setError } from '../../../Store/UI/UIActions';

jest.mock('../../../Store/Auth/AuthActions', () => ({
    startRegister: jest.fn(),
    startLogin: jest.fn()
}));

jest.mock('../../../Store/UI/UIActions', () => ({
    setError: jest.fn(),
    rmError: jest.fn()
}));

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const initialState = {
    UI: {
        msg: null
    }
};
let store = mockStore( initialState );
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <LoginScreen />
    </Provider>
)

describe('Pruebas en <LoginScreen />', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should first', () => {
        expect( wrapper ).toMatchSnapshot();
    });
    
    test('debe de llamar el dispatch del login', () => {
        wrapper.find('form').at(0).simulate('submit', { preventDefault(){} });
        wrapper.find('input[name="lEmail"]').simulate('change', {
            target:{
                name: 'lEmail',
                value: 'jonathan@jojo.com'
            }
        });
        wrapper.find('input[name="lPassword"]').simulate('change', {
            target:{
                name: 'rPassword',
                value: 'jjba123'
            }
        });
        expect( setError ).not.toHaveBeenCalled();
        expect( rmError ).toHaveBeenCalled();
        expect( startLogin ).toHaveBeenCalledWith('jonathan@jojo.com', 'jjba123');
    });

    test('No hay registro si las contraseña son difernetes', () => {
        wrapper.find('input[name="rPassword"]').simulate('change', {
            target:{
                name: 'rPassword',
                value: 'jojoba1'
            }
        });

        wrapper.find('input[name="rPasswordConfirm"]').simulate('change', {
            target: {
                name: 'rPasswordConfirm',
                value: 'jojoba2'
            }
        });

        wrapper.find('form').at(1).simulate('submit', { preventDefault(){} });

        expect( setError ).toHaveBeenCalledWith('Password and Confirm Password should be equals');
        expect( rmError ).not.toHaveBeenCalled();
        expect( startRegister ).not.toHaveBeenCalled();
    });

    test('Registro con contraseñas iguales', () => {
        wrapper.find('input[name="rName"]').simulate('change', {
            target:{
                name: 'rName',
                value: 'jonathan'
            }
        });
        
        wrapper.find('input[name="rEmail"]').simulate('change', {
            target:{
                name: 'rEmail',
                value: 'jonathan@jojo.com'
            }
        });

        wrapper.find('input[name="rPassword"]').simulate('change', {
            target:{
                name: 'rPassword',
                value: 'jjba123'
            }
        });
        
        wrapper.find('input[name="rPasswordConfirm"]').simulate('change', {
            target: {
                name: 'rPasswordConfirm',
                value: 'jjba123'
            }
        });

        wrapper.find('form').at(1).simulate('submit', { preventDefault(){} });
        expect( setError ).not.toHaveBeenCalled();
        expect( rmError ).toHaveBeenCalled();
        expect( startRegister ).toHaveBeenCalledWith('jonathan@jojo.com', 'jjba123', 'jonathan');
    });
    
})