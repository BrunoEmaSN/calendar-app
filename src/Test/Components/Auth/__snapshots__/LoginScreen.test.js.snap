// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> should first 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Correo"
            required={true}
            type="text"
            value="jonathan@jojo.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            required={true}
            type="password"
            value="jjba123"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            required={true}
            type="text"
            value="Joseph"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Correo"
            required={true}
            type="email"
            value="joseph@jojo.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            required={true}
            type="password"
            value="jjba123"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPasswordConfirm"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            required={true}
            type="password"
            value="jjba123"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
